Basic JWT implementation

cpiber opened the pull request:
As discussed on discord

Instead of the transparent server-side login, switch to a JWT implementation. This requires cooperation from the UI.

Very heavily based on #623

cpiber committed:
Basic JWT implementation

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt:
```diff
@@ -0,0 +1,97 @@
+        val secretKey = SecretKeySpec(keyBytes, ALGORITHM)
+
+        return Base64.encode(secretKey.encoded)
+    }
```
This key needs to be generated only once, and then save it to be reused. We can save it into a SharedPreferences like we do in TrackerPreferences

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt:
```diff
@@ -0,0 +1,97 @@
+    private val accessTokenExpiry = 1.hours
+    private val refreshTokenExpiry = 60.days
+    private const val ISSUER = "tachidesk"
+    private const val AUDIENCE = "" // todo audience
```
This TODO should probably be investigated(I had not gotten to it yet)

cpiber replied:
From a quick search, this should usually be the public url of the API, but this is not really known here. We could make it a configurable setting? Another option is also just some generic string like "tenant-xyz" or "service-A", so the setting could really be any string, maybe defaulting to "suwayomi-server".

https://mojoauth.com/blog/lets-understand-jwt-audience-aud-claim/

Syer10 replied:
That works for me

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt:
```diff
@@ -0,0 +1,97 @@
+    private const val ALGORITHM = "HmacSHA256"
+    private val accessTokenExpiry = 1.hours
+    private val refreshTokenExpiry = 60.days
+    private const val ISSUER = "tachidesk"
```
```suggestion
    private const val ISSUER = "suwayomi-server"
```

cpiber committed:
Move JWT to UI_LOGIN mode and bring back SIMPLE_LOGIN as before

cpiber committed:
Update server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt

Co-authored-by: Mitchell Syer <Syer10@users.noreply.github.com>

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt:
```diff
@@ -0,0 +1,97 @@
+
+    private const val ALGORITHM = "HmacSHA256"
+    private val accessTokenExpiry = 1.hours
+    private val refreshTokenExpiry = 60.days
```
Maybe make it user-configurable?

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt:
```diff
@@ -0,0 +1,97 @@
+        require(jwt.getClaim("token_type").asString() == "refresh") {
+            "Cannot use access token to refresh"
+        }
+        return generateJwt()
```
```suggestion
    fun refreshJwt(refreshToken: String): String {
        val jwt = verifier.verify(refreshToken)
        require(jwt.getClaim("token_type").asString() == "refresh") {
            "Cannot use access token to refresh"
        }
        return createAccessToken()
```
Lets only refresh the access token, I want the refresh token to only be given once per login.
You will need to change the refreshToken mutation as well

cpiber replied:
Makes sense, I think that's the typical usecase

cpiber committed:
Refresh: Update only access token

Co-authored-by: Mitchell Syer <Syer10@users.noreply.github.com>

cpiber committed:
Implement JWT Audience

cpiber committed:
Store JWT key

Generates the key on startup if not set

cpiber committed:
Handle invalid Base64

cpiber committed:
Make JWT expiry configurable

cpiber committed:
Missing value parse

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/server/ServerConfig.kt:
```diff
@@ -69,6 +70,33 @@ class ServerConfig(
+            return stateFlow
+        }
+    }
+
```
You don't actually need this, you can register a custom Config type like I have done with MutableStateFlowType(search it and its usages), just simpler since its a basic type.

cpiber replied:
But we still want to have a MutableStateFlow, right? So it would need to match a `MutableStateFlow<Duration>`. I'll look into that tomorrow

Syer10 replied:
Nope, it just needs to match Duration, the way I have MutableStateFlow is just a wrapper for the internal type deserialization. So you only need to register a Duration custom type

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt:
```diff
@@ -0,0 +1,126 @@
+        return Base64.encode(secretKey.encoded)
+    }
+
+    private val algorithm: Algorithm = Algorithm.HMAC256(generateSecret()) // todo store secret
```
```suggestion
    private val algorithm: Algorithm = Algorithm.HMAC256(generateSecret())
```

cpiber committed:
Update server/src/main/kotlin/suwayomi/tachidesk/global/impl/util/Jwt.kt

Co-authored-by: Mitchell Syer <Syer10@users.noreply.github.com>

cpiber committed:
Simplify Duration parsing

cpiber committed:
JWT Protect Mutations

cpiber committed:
JWT Protect Queries and Subscriptions

cpiber committed:
JWT Protect v1 WebSockets

cpiber committed:
WebSockets allow sending token via protocol header

cpiber committed:
Also respect the `suwayomi-server-token` cookie

cpiber committed:
JWT reduce default token expiry

cpiber committed:
JWT Support cookie on WebSocket as well

cpiber committed:
Lint

schroda committed:
Authenticate graphql subscription via connection_init payload

cpiber committed:
Merge branch 'master' into login-api

cpiber committed:
Merge branch 'master' into login-api

cpiber committed:
WebView: Prefer explicit token over cookie

This hack was implemented because WebView sent `"null"` if no token was
supplied, just don't send a bad token, then we can do this properly

cpiber committed:
WebView: Implement basic login dialog if no token supplied

Syer10 reviewed (commented):
Was the basic auth alternative for the OPDS endpoints implememented? I don't see it.

Syer10 committed:
Merge branch 'master' into login-api
